pandas/tseries/index.py

```diff
@@ -681,6 +687,67 @@ def _add_delta(self, delta):
             result = result.tz_convert(self.tz)
         return result
 
+    def _add_offset(self, offset):
+        from pandas import Timedelta
+        i = self
+        n = offset.n
+
+        def _begend_offset(i, n, offset, anchor='beg'):
+            n += 0 if n >= 0 else 1
+            base = (i.to_period(offset) + n).to_timestamp()
+            if anchor == 'end':
+                base = (base.to_period('D') - 1).to_timestamp()
+            off = i.period_delta('D')
+            return base + off
+
+        # fastpasth a subset of relativedelta
+        relativedelta_fast = {'years', 'months', 'weeks',
+                              'days', 'hours', 'minutes',
+                              'seconds', 'microseconds'}
+
+        if type(offset) is DateOffset:
+            if offset._use_relativedelta:
+                if set(offset.kwds).issubset(relativedelta_fast):
+                    months = ((offset.kwds.get('years', 0) * 12
+                              + offset.kwds.get('months', 0)) * n)
+                    if months:
+                        base = (i.to_period('M') + months).to_timestamp()
+                        # minimum prevent month-end from wrapping
+                        off = np.minimum(i.period_delta('M'),
+                                         to_timedelta(base.days_in_month - 1, unit='D'))
+                        i = base + off
+
+                    weeks = (offset.kwds.get('weeks', 0)) * n
+                    if weeks:
+                        i = (i.to_period('W') + weeks).to_timestamp() + i.period_delta('W')
+
+                    d = (offset.kwds.get('days', 0)) * n
```

Comment:
you can do this in the constructor of the ``Timedelta`` itself, e.g. something like

delta = ``Timedelta(**offset.kwds)``